routers/api/v1/repo/search.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package repo
+
+import (
+	"net/http"
+
+	"code.gitea.io/gitea/modules/context"
+	code_indexer "code.gitea.io/gitea/modules/indexer/code"
+	"code.gitea.io/gitea/modules/setting"
+	api "code.gitea.io/gitea/modules/structs"
+	"code.gitea.io/gitea/routers/api/v1/utils"
+	"code.gitea.io/gitea/services/convert"
+)
+
+// CodeSearch Performs a code search on a Repo
+func CodeSearch(ctx *context.APIContext) {
+	// swagger:operation GET /repos/{owner}/{repo}/code_search repository repoCodeSearch
+	// ---
+	// summary: Performs a code search on a Repo
+	// produces:
+	// - application/json
+	// parameters:
+	// - name: owner
+	//   in: path
+	//   description: owner of the repo
+	//   type: string
+	//   required: true
+	// - name: repo
+	//   in: path
+	//   description: name of the repo
+	//   type: string
+	//   required: true
+	// - name: keyword
+	//   in: query
+	//   description: the keyword the search for
+	//   type: string
+	// - name: language
+	//   in: query
+	//   description: filter results by language
+	//   type: string
+	// - name: match
+	//   in: query
+	//   description: only exact match (defaults to true)
+	//   type: boolean
+	// - name: page
+	//   in: query
+	//   description: page number of results to return (1-based)
+	//   type: integer
+	// - name: limit
+	//   in: query
+	//   description: page size of results
+	//   type: string
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/RepoCodeSearch"
+	//   "422":
+	//     description: "The keyword is empty"
+	//     schema:
+	//       type: string
+	if !setting.Indexer.RepoIndexerEnabled {
+		ctx.Error(http.StatusInternalServerError, "IndexerNotEnabled", "The Code Indexer is not enabled on this server")
```

Comment:
Yes, 501 seems better here. Imho 500 should be reserved for unexpected errors.

Reply:
I don't think 5xx status codes would be correct in such cases. I would say if the feature is disabled and because of that API is not available it should be 4xx status code, I would say 404

Reply:
5xx means the problem is on the server side. 4xx on the client side. If code search is not enabled, the problem is on the server side.

Reply:
Problem is really on client side for calling API that is disabled (not there) thus 404 just like 403 for calling method you don't have rights

Reply:
5xx is for status codes where the server knows it should be able to complete but for unexpected reasons it cannot but it in this case server is not willing to complete the request because it's not just not possible thus it should be 4xx status code